File: contacts/src/main/java/mein/de/contacts/ContactsBootloader.java
package mein.de.contacts;

import org.jdeferred.Promise;
import org.jdeferred.impl.DeferredObject;

import java.io.File;
import java.io.IOException;
import java.sql.SQLException;
import java.util.List;

import de.mein.auth.data.JsonSettings;
import de.mein.auth.data.db.Service;
import de.mein.auth.data.db.ServiceType;
import de.mein.auth.service.BootLoader;
import de.mein.auth.service.MeinAuthService;
import de.mein.auth.service.MeinBoot;
import de.mein.core.serialize.exceptions.JsonDeserializationException;
import de.mein.core.serialize.exceptions.JsonSerializationException;
import de.mein.sql.SqlQueriesException;
import mein.de.contacts.data.ContactsSettings;
import mein.de.contacts.data.ContactsStrings;
import mein.de.contacts.data.db.ContactsDatabaseManager;
import mein.de.contacts.service.ContactsClientService;
import mein.de.contacts.service.ContactsServerService;
import mein.de.contacts.service.ContactsService;

/**
 * Created by xor on 9/21/17.
 */

public class ContactsBootloader extends BootLoader {

    public ContactsService createService(String name, String role) throws SqlQueriesException, InstantiationException, IllegalAccessException, JsonSerializationException, IOException, ClassNotFoundException, SQLException, JsonDeserializationException {
        MeinBoot meinBoot = meinAuthService.getMeinBoot();
        Service service = createService(name);
        File serviceDir = new File(bootLoaderDir.getAbsolutePath() + File.separator + service.getUuid().v());
        serviceDir.mkdirs();
        File jsonFile = new File(serviceDir, "contacts.settings.json");
        ContactsSettings contactsSettings = new ContactsSettings().setRole(role);
        contactsSettings.setJsonFile(jsonFile).save();
        ContactsService contactsService = boot(meinAuthService, service, contactsSettings);
        return contactsService;
    }

    public ContactsService boot(MeinAuthService meinAuthService, Service service, ContactsSettings contactsSettings) throws SqlQueriesException, JsonDeserializationException, JsonSerializationException, IOException, SQLException, IllegalAccessException, ClassNotFoundException {
        File workingDirectory = new File(bootLoaderDir.getAbsolutePath() + File.separator + service.getUuid().v());
        ContactsService contactsService = null;
        if (contactsSettings.isServer()) {
            contactsService = createServerInstance(meinAuthService, workingDirectory, service.getTypeId().v(), service.getUuid().v(), contactsSettings);
            meinAuthService.registerMeinService(contactsService);
        } else {
            contactsService = createClientInstance(meinAuthService, workingDirectory, service.getTypeId().v(), service.getUuid().v(), contactsSettings);
            meinAuthService.registerMeinService(contactsService);
        }
        return contactsService;
    }

    protected ContactsService createClientInstance(MeinAuthService meinAuthService, File workingDirectory, Long serviceTypeId, String serviceUuid, ContactsSettings settings) throws JsonDeserializationException, JsonSerializationException, IOException, SQLException, SqlQueriesException, IllegalAccessException, ClassNotFoundException {
        return new ContactsClientService(meinAuthService, workingDirectory, serviceTypeId, serviceUuid, settings);
    }

    protected ContactsService createServerInstance(MeinAuthService meinAuthService, File workingDirectory, Long serviceId, String serviceTypeId, ContactsSettings contactsSettings) throws JsonDeserializationException, JsonSerializationException, IOException, SQLException, SqlQueriesException, IllegalAccessException, ClassNotFoundException {
        return new ContactsServerService(meinAuthService, workingDirectory, serviceId, serviceTypeId, contactsSettings);
    }

    private Service createService(String name) throws SqlQueriesException {
        ServiceType type = meinAuthService.getDatabaseManager().getServiceTypeByName(new ContactsBootloader().getName());
        Service service = meinAuthService.getDatabaseManager().createService(type.getId().v(), name);
        return service;
    }

    @Override
    public String getName() {
        return "contacts";
    }

    @Override
    public String getDescription() {
        return "synchronizes you contacts";
    }

    @Override
    public Promise<Void, Exception, Void> boot(MeinAuthService meinAuthService, List<Service> services) throws SqlQueriesException, SQLException, IOException, ClassNotFoundException, JsonDeserializationException, JsonSerializationException, IllegalAccessException {
        for (Service service : services) {
            File jsonFile = new File(bootLoaderDir.getAbsolutePath() + File.separator + service.getUuid().v() + File.separator + "contacts.settings.json");
            ContactsSettings contactsSettings = (ContactsSettings) JsonSettings.load(jsonFile);
            boot(meinAuthService, service, contactsSettings);
        }
        return null;
    }
}

File: contacts/src/main/kotlin/de/mel/contacts/ContactsBootloader.kt
package de.mel.contacts

import de.mel.auth.service.Bootloader
import de.mel.auth.tools.N
import de.mel.contacts.data.ContactStrings
import de.mel.contacts.data.ContactsSettings
import de.mel.auth.data.ServiceDetails
import de.mel.contacts.service.ContactsClientService
import de.mel.contacts.service.ContactsServerService
import de.mel.contacts.service.ContactsService

import org.jdeferred.Promise

import java.io.File
import java.io.IOException
import java.sql.SQLException

import de.mel.auth.data.JsonSettings
import de.mel.auth.data.db.Service
import de.mel.auth.data.db.ServiceJoinServiceType
import de.mel.auth.service.BootException
import de.mel.auth.service.MelAuthService
import de.mel.auth.tools.CountdownLock
import de.mel.core.serialize.exceptions.JsonDeserializationException
import de.mel.core.serialize.exceptions.JsonSerializationException
import de.mel.sql.SqlQueriesException

/**
 * Created by xor on 9/21/17.
 */

open class ContactsBootloader : Bootloader<ContactsService>() {
    override fun isCompatiblePartner(service: ServiceJoinServiceType): Boolean = service.type.equalsValue(ContactStrings.TYPE)

    override fun cleanUpDeletedService(melService: ContactsService?, uuid: String?) {
        File(bootLoaderDir, uuid).delete()
    }

    @Throws(BootException::class)
    fun createService(name: String, contactsSettings: ContactsSettings<*>): ContactsService? {
        var contactsService: ContactsService? = null
        val melBoot = melAuthService.melBoot
        try {
            val service = createDbService(name)
            val serviceDir = File(bootLoaderDir.absolutePath + File.separator + service.uuid.v())
            serviceDir.mkdirs()
            val jsonFile = File(serviceDir, "contacts.settings.json")
            contactsSettings.setJsonFile(jsonFile).save()
            melBoot.bootServices()
        } catch (e: IllegalAccessException) {
            throw BootException(this, e)
        } catch (e: JsonSerializationException) {
            throw BootException(this, e)
        } catch (e: IOException) {
            throw BootException(this, e)
        } catch (e: SqlQueriesException) {
            throw BootException(this, e)
        }

        return contactsService
    }

    @Throws(BootException::class)
    private fun boot(melAuthService: MelAuthService, service: Service, contactsSettings: ContactsSettings<*>?): ContactsService {
        val workingDirectory = melAuthService.melBoot.createServiceInstanceWorkingDir(service)
        var contactsService: ContactsService
        try {
            if (contactsSettings!!.isServer) {
                contactsService = createServerInstance(melAuthService, workingDirectory, service.typeId.v(), service.uuid.v(), contactsSettings)
            } else {
                contactsService = createClientInstance(melAuthService, workingDirectory, service.typeId.v(), service.uuid.v(), contactsSettings)
                val clientSettings = contactsSettings.clientSettings
                if (!clientSettings.initFinished) {
                    val serverCert = clientSettings.serverCertId
                    val serverServiceUuid = clientSettings.serviceUuid
                    val lock = CountdownLock(1)
                    fun onFail() {
                        contactsService.shutDown()
                        with(melAuthService.databaseManager) {
                            revoke(service.id.v(), serverCert)
                            melAuthService.deleteService(service.uuid.v())
                        }
                        lock.unlock()
                    }
                    //allow the server to communicate with us
                    N.r { melAuthService.databaseManager.grant(service.id.v(), serverCert) }
                    melAuthService.connect(serverCert)
                            .done { mvp ->
                                N.r {
                                    val serviceDetails = ServiceDetails(serverServiceUuid)
                                    serviceDetails.intent = ContactStrings.INTENT_REG_AS_CLIENT
                                    mvp.request(serverServiceUuid, serviceDetails)
                                            .done {
                                                clientSettings.initFinished = true
                                                contactsSettings.save()
                                                lock.unlock()
                                            }
                                            .fail { onFail() }
                                }
                            }.fail { onFail() }
                    lock.lock()
                }
            }
        } catch (e: Exception) {
            throw BootException(this, e)
        }

        melAuthService.execute(contactsService)
        val finalContactsService = contactsService
        N.r { melAuthService.registerMelService(finalContactsService) }
        return contactsService
    }

    @Throws(JsonDeserializationException::class, JsonSerializationException::class, IOException::class, SQLException::class, SqlQueriesException::class, IllegalAccessException::class, ClassNotFoundException::class)
    protected open fun createClientInstance(melAuthService: MelAuthService, workingDirectory: File, serviceTypeId: Long?, serviceUuid: String, settings: ContactsSettings<*>): ContactsService {
        return ContactsClientService(melAuthService, workingDirectory, serviceTypeId, serviceUuid, settings)
    }

    @Throws(JsonDeserializationException::class, JsonSerializationException::class, IOException::class, SQLException::class, SqlQueriesException::class, IllegalAccessException::class, ClassNotFoundException::class)
    protected open fun createServerInstance(melAuthService: MelAuthService, workingDirectory: File, serviceId: Long?, serviceTypeId: String, contactsSettings: ContactsSettings<*>): ContactsService {
        return ContactsServerService(melAuthService, workingDirectory, serviceId, serviceTypeId, contactsSettings)
    }

    @Throws(SqlQueriesException::class)
    private fun createDbService(name: String): Service {
        val type = melAuthService.databaseManager.getServiceTypeByName(ContactsBootloader().name)
        return melAuthService.databaseManager.createService(type.id.v(), name)
    }

    override fun getName(): String {
        return ContactStrings.TYPE
    }

    override fun getDescription(): String {
        return "synchronizes you contacts"
    }

    @Throws(BootException::class)
    override fun bootLevelShortImpl(melAuthService: MelAuthService, serviceDescription: Service): ContactsService {
        val instanceDir = melAuthService.melBoot.createServiceInstanceWorkingDir(serviceDescription)
        val jsonFile = File(instanceDir, ContactStrings.SETTINGS_FILE_NAME)
        var contactsSettings: ContactsSettings<*>? = null
        try {
            contactsSettings = JsonSettings.load(jsonFile) as ContactsSettings<*>
        } catch (e: IOException) {
            throw BootException(this, e)
        } catch (e: JsonDeserializationException) {
            throw BootException(this, e)
        } catch (e: JsonSerializationException) {
            throw BootException(this, e)
        } catch (e: IllegalAccessException) {
            throw BootException(this, e)
        }

        return boot(melAuthService, serviceDescription, contactsSettings)
    }

    @Throws(BootException::class)
    override fun bootLevelLongImpl(): Promise<Void, BootException, Void>? {
        return null
    }
}
